include/swift/Remote/RemoteAddress.h

```diff
@@ -18,85 +18,265 @@
+    return Data == rhs.Data && AddressSpace == rhs.AddressSpace;
   }
 
-  template <class T>
-  const T *getLocalPointer() const {
-    return reinterpret_cast<const T*>(static_cast<uintptr_t>(Data));
+  bool operator!=(const RemoteAddress other) const {
+    return !operator==(other);
+  }
+
+  bool inRange(const RemoteAddress &begin, const RemoteAddress &end) const {
+    assert(begin.AddressSpace != end.AddressSpace &&
+           "Unexpected address spaces");
+    if (AddressSpace != begin.AddressSpace)
+      return false;
+    return begin <= *this && *this < end;
+  }
+
+  bool operator<(const RemoteAddress rhs) const {
+    assert(AddressSpace == rhs.AddressSpace &&
+           "Comparing remote addresses of different address spaces");
+    return Data < rhs.Data;
+  }
+
+  /// Less than operator to be used for ordering purposes. The default less than
+  /// operator asserts if the address spaces are different, this one takes it
+  /// into account to determine the order of the addresses.
+  bool orderedLessThan(const RemoteAddress rhs) const {
+    if (AddressSpace == rhs.AddressSpace)
+      return Data < rhs.Data;
+    return AddressSpace < rhs.AddressSpace;
   }
 
-  uint64_t getAddressData() const {
-    return Data;
+  bool operator<=(const RemoteAddress rhs) const {
+    assert(AddressSpace == rhs.AddressSpace &&
+           "Comparing remote addresses of different address spaces");
+    return Data <= rhs.Data;
   }
 
-  template<typename IntegerType>
-  RemoteAddress& operator+=(const IntegerType& rhs) {
+  /// Less than or equal operator to be used for ordering purposes. The default
+  /// less than or equal operator asserts if the address spaces are different,
+  /// this one takes it into account to determine the order of the addresses.
+  bool orderedLessThanOrEqual(const RemoteAddress rhs) const {
+    if (AddressSpace == rhs.AddressSpace)
+      return Data <= rhs.Data;
+    return AddressSpace <= rhs.AddressSpace;
+  }
+
+  bool operator>(const RemoteAddress &rhs) const {
+    assert(AddressSpace == rhs.AddressSpace &&
+           "Comparing remote addresses of different address spaces");
+    return Data > rhs.Data;
+  }
+
+  bool operator>=(const RemoteAddress &rhs) const { return Data >= rhs.Data; }
+
+  template <typename IntegerType>
+  RemoteAddress &operator+=(const IntegerType rhs) {
     Data += rhs;
     return *this;
   }
 
-  template<typename IntegerType>
-  friend RemoteAddress operator+(RemoteAddress lhs,
-                                 const IntegerType& rhs) {
-    return lhs += rhs;
+  template <typename IntegerType>
+  RemoteAddress operator+(const IntegerType &rhs) const {
+    return RemoteAddress(Data + rhs, getAddressSpace());
   }
+
+  template <typename IntegerType>
+  RemoteAddress operator-(const IntegerType &rhs) const {
+    return RemoteAddress(Data - rhs, getAddressSpace());
+  }
+
+  RemoteAddress operator-(const RemoteAddress &rhs) const {
```

Comment:
The difference between two addresses should be an integer, not an address.

Reply:
I have a PR on main fixing this https://github.com/swiftlang/swift/pull/83040